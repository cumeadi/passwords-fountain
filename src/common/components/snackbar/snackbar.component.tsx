import { h } from 'preact';
import { TypedComponent } from '@/common/typings/prop-types';
import PropTypes from 'prop-types';
import { forbidExtraProps } from 'airbnb-prop-types';
import {
    Wrapper,
    IconWrapper,
    TextWrapper,
    ErrorType,
} from './snackbar.styles';
import { useSelector } from '@preact-hooks/unistore';
import {
    selectSnackbarMessageKey,
    selectSnackbarType,
} from '@/modules/overlay/overlay.selectors';
import { SnackbarType } from '@/modules/overlay/overlay.constants';
import { Icon } from '../icon';
import { Text } from '@/modules/localisation/localisation.context';

export const Snackbar: TypedComponent<Props> = () => {
    const messageKey = useSelector(selectSnackbarMessageKey);
    const type = useSelector(selectSnackbarType) as SnackbarType;
    const iconSize = 44;

    return (
        <Wrapper type={type}>
            <IconWrapper>
                <Icon name={type} width={iconSize} height={iconSize} />
            </IconWrapper>
            <TextWrapper>
                <ErrorType>
                    <Text>{type}</Text>
                </ErrorType>
                <span>{' - '}</span>
                <Text>{messageKey}</Text>
            </TextWrapper>
        </Wrapper>
    );
};

interface Props {}
Snackbar.propTypes = forbidExtraProps({});
