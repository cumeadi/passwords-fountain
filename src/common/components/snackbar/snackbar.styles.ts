import { h } from 'preact';
import { styled, setPragma } from 'goober';
import { theme, typography, stylelint } from '@/common/theme';
import {
    snackbarTypes,
    SnackbarType,
} from '@/modules/overlay/overlay.constants';
setPragma(h);

const info = stylelint.css`
    border: 1px solid ${theme.colors.primaryBlue};
    border-top: 4px solid ${theme.colors.primaryBlue};
`;

const success = stylelint.css`
    border: 1px solid ${theme.colors.primaryGreen};
    border-top: 4px solid ${theme.colors.primaryGreen};
`;

const error = stylelint.css`
    border: 1px solid ${theme.colors.red};
    border-top: 4px solid ${theme.colors.red};
`;

const getVariantStyles = ({ type }: { type: SnackbarType }): string => {
    switch (type) {
        case snackbarTypes.info:
            return info;
        case snackbarTypes.success:
            return success;
        case snackbarTypes.error:
            return error;
    }

    return info;
};

export const Wrapper = styled('section')`
    ${getVariantStyles}
    ${typography.text18}
    max-width: 500px;
    width: 100%;
    position: absolute;
    bottom: 2px;
    left: 50%;
    transform: translateX(-50%);
    padding: ${theme.spacing.s12};
    background: ${theme.colors.white};
    color: ${theme.colors.darkBlue};
`;
