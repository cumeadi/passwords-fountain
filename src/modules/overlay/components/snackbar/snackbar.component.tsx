import { h } from 'preact';
import { TypedComponent } from '@/common/typings/prop-types';
import {
    Wrapper,
    IconWrapper,
    TextWrapper,
    ErrorType,
} from './snackbar.styles';
import { useSelector, useAction } from '@/store';
import { snackbarVisibilityTime } from '@/modules/overlay/overlay.constants';
import { selectSnackbarMessages } from '@/modules/overlay/overlay.selectors';
import { overlayActions } from '@/modules/overlay/overlay.actions';
import { Icon } from '@/common/components/icon';
import { Text } from '@/modules/localisation/components/text';
import { useEffect } from 'preact/hooks';

export const Snackbar: TypedComponent<Props> = () => {
    const iconSize = 44;
    const [currentMessage] = useSelector(selectSnackbarMessages);
    const hideSnackbar = useAction(overlayActions.hideSnackbar);

    useEffect(() => {
        setTimeout(() => {
            hideSnackbar(currentMessage.id);
        }, snackbarVisibilityTime);
    }, [currentMessage]);

    return (
        <Wrapper type={currentMessage.type}>
            <IconWrapper>
                <Icon
                    name={currentMessage.type}
                    width={iconSize}
                    height={iconSize}
                />
            </IconWrapper>
            <TextWrapper>
                <ErrorType>
                    <Text>{currentMessage.type}</Text>
                </ErrorType>
                <span>{' - '}</span>
                <Text>{currentMessage.messageKey}</Text>
            </TextWrapper>
        </Wrapper>
    );
};

interface Props {}
Snackbar.propTypes = {};
