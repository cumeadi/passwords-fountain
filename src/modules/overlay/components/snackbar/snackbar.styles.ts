import { h } from 'preact';
import { styled, setPragma } from 'goober';
import { theme, typography, stylelint } from '@/common/theme';
import {
    snackbarTypes,
    SnackbarType,
} from '@/modules/overlay/overlay.constants';
setPragma(h);

const info = stylelint.css`
    border: 1px solid ${theme.colors.primaryBlue};
    border-top: 4px solid ${theme.colors.primaryBlue};
`;

const success = stylelint.css`
    border: 1px solid ${theme.colors.primaryGreen};
    border-top: 4px solid ${theme.colors.primaryGreen};
`;

const error = stylelint.css`
    border: 1px solid ${theme.colors.red};
    border-top: 4px solid ${theme.colors.red};
`;

const getVariantStyles = ({ type }: { type: SnackbarType }): string => {
    switch (type) {
        case snackbarTypes.info:
            return info;
        case snackbarTypes.success:
            return success;
        case snackbarTypes.error:
            return error;
    }

    return info;
};

export const Wrapper = styled('section')`
    ${getVariantStyles}
    z-index: ${theme.zIndexes.s};
    display: grid;
    grid-template-columns: auto 1fr;
    grid-gap: ${theme.spacing.s12};
    max-width: 500px;
    width: 100%;
    position: fixed;
    bottom: ${theme.spacing.xxs2};
    left: 50%;
    transform: translateX(-50%);
    padding: ${theme.spacing.s12};
    background: ${theme.colors.white};
    color: ${theme.colors.darkBlue};
`;

export const IconWrapper = styled('div')``;
export const TextWrapper = styled('div')`
    ${typography.text18}
`;

export const ErrorType = styled('strong')`
    font-family: ${theme.fontFamilies.bold};
`;
