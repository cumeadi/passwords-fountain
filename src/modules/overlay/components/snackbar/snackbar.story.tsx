import { h, VNode } from 'preact';
import { Snackbar } from './snackbar.component';
import { withA11y } from '@storybook/addon-a11y';
import { withStoreMock } from '../../../../../.storybook/helpers';

export default {
    title: 'Snackbar',
    decorators: [withA11y],
};

const successData = {
    id: 0,
    messageKey: 'passwords fetched correctly!',
    type: 'success',
};

const infoData = {
    id: 0,
    messageKey: 'You are offline.',
    type: 'info',
};

const errorData = {
    id: 0,
    messageKey: 'cannot connect to your database.',
    type: 'error',
};

const getState = (snackbarPayload: any) => ({
    overlay: {
        snackbarMessages: [snackbarPayload],
    },
});

export const defaultView = (): VNode =>
    withStoreMock(() => <Snackbar />, getState(infoData));

export const success = (): VNode =>
    withStoreMock(() => <Snackbar />, getState(successData));

export const error = (): VNode =>
    withStoreMock(() => <Snackbar />, getState(errorData));
