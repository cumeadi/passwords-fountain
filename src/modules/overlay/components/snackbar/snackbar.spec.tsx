import '@testing-library/jest-dom/extend-expect';
import { h } from 'preact';
import { render } from '@testing-library/preact';
import { Snackbar } from './snackbar.component';
import { StoreProvider } from '@preact-hooks/unistore';
import createStore from 'unistore';

const stateMock = {
    overlay: {
        snackbarMessages: [
            {
                id: 0,
                type: 'success',
                messageKey: 'Passwords fetched!',
            },
        ],
    },
};

describe('Snackbar', () => {
    it('should render correctly', () => {
        const { asFragment } = render(
            <StoreProvider value={createStore(stateMock)}>
                <Snackbar />
            </StoreProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
