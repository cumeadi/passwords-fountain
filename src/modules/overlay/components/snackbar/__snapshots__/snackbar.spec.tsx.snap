// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar should render correctly 1`] = `
<DocumentFragment>
  <section
    class="go2645182152"
    type="success"
  >
    <div
      class="go115679"
    >
      <img
        alt="icon"
        class="go115679"
        height="44"
        src="static-asset-stub"
        width="44"
      />
    </div>
    <div
      class="go2656262278"
    >
      <strong
        class="go859468004"
      >
        Success
      </strong>
      <span>
         - 
      </span>
      Passwords fetched!
    </div>
  </section>
</DocumentFragment>
`;
